---
title: "Differential Expression RNAseq Mouse Kidney"
author: "Uwe Schwartz"
date: "02/12/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_knit$set(root.dir = "~/Analysis/17_20191202_Mm_RNAseq//")
setwd("~/Analysis/17_20191202_Mm_RNAseq/")
out.path<-"~/Analysis/17_20191202_Mm_RNAseq//analysis/"
date.x<-Sys.Date()
```

## Dependencies & Libraries

#### Load packages
```{r libraries, message=FALSE}
library(DESeq2)
library(RColorBrewer)
library(ggplot2)
library("pheatmap")
palette(unique(c(brewer.pal(9,"Set1"),brewer.pal(8,"Set2"))))
```

##Import data

First load DESeq2 object generated in exploratory ananylsis. Then perform differential gene expression analysis based on KO versus WT mice

```{r DESeq obj}
# load 
load(file=paste(out.path,"obj/dds.obj.rda",sep="/"))

# meta data
colData(dds.obj)

#design
design(dds.obj)<-~condition

## differential gene expression
dds.diff<-DESeq(dds.obj)


### get annotation of genes
biotype<-read.delim("~/Annotation/MusMusculus/protein_coding_and_lincRNA_INFO.txt")

mx<-match(rownames(rowData(dds.diff)),biotype$ensembl_gene_id)
rowData(dds.diff)$biotype<-biotype$gene_biotype[mx]
rowData(dds.diff)$symbol<-biotype$mgi_symbol[mx]
rowData(dds.diff)$EnsemblID<-biotype$ensembl_gene_id[mx]
```


```{r checkpoint1, eval=FALSE, echo=FALSE }
save(dds.diff, file=paste(out.path,"obj/dds.diff.rda",sep="/"))
```

## WT versus KO

```{r KO vs WT}
#### differential analysis
ctrl="WT"
cond="KO"

# get results
res<-results(dds.diff, contrast=c("condition",cond, ctrl))

summary(res)

#MA plot before shrinkage
max.lfc<-max(abs(res$log2FoldChange),na.rm = T)

DESeq2::plotMA(res,ylim = c(-max.lfc, max.lfc), cex=0.5)  

#shrinkage
res.shr <- lfcShrink(dds.diff, 
                     contrast=c("condition",cond, ctrl), 
                     res=res, type = "ashr") 

max.lfc<-max(abs(res.shr$log2FoldChange),na.rm = T)

DESeq2::plotMA(res.shr,ylim = c(-max.lfc, max.lfc), cex=0.5)
    abline(h=c(-1,1), lty=2)

    
## pVal_distribution

hist(res$pvalue[res$baseMean > 1], breaks = 0:20/20,
       col = "grey50", border = "white", xlab="p-Value", main="pVal distribution")
     

# get Gene annotation
res<-cbind(res.shr,rowData(dds.diff)[rownames(res.shr),
                                    c("EnsemblID","biotype","symbol")])
    

head(res[order(res$padj),])



### significant hits
res.sig<-subset(res, padj<0.05 & abs(res$log2FoldChange)>=1)

#significant hits up regulated
table(res.sig$log2FoldChange>0)

## heatmap of significant genes
load(file=paste(out.path,"obj/rld.rda",sep="/"))

mat<-assay(rld)[rownames(res.sig),]
rownames(mat)<- res.sig$symbol
anno<-as.data.frame(colData(dds.diff)[,c("condition", "replica")])
  
print(pheatmap(mat,scale="row", annotation_col = anno,show_rownames  = F) )


```


```{r diff exp sh3 save data, eval=FALSE, echo=FALSE }
dir.create(paste(out.path,"diffExp",sep="/"))

path.diff.sh3=paste(out.path,"diffExp",sep="/")
dir.create(path.diff.sh3, showWarnings = F)

  
pdf(paste0(path.diff.sh3,"/MA_Shrinkage.pdf"), width=5,height = 5)
    DESeq2::plotMA(res.shr,ylim = c(-max.lfc, max.lfc), cex=0.5)
    abline(h=c(-1,1), lty=2)
dev.off()

## pVal_distribution
pdf(paste0(path.diff.sh3,"/pVal_histogram.pdf"), width=5,height = 5)
    hist(res$pvalue[res$baseMean > 1], breaks = 0:20/20,
       col = "grey50", border = "white", xlab="p-Value", main="pVal distribution")
dev.off()

save(res, file=paste0(path.diff.sh3,"/res.rda"))
write.table(res[order(res$padj),], file=paste0(path.diff.sh3,"/res_table.txt"),
            row.names = F,quote=F, sep="\t")


############### glimma ## eploratory
  library(Glimma)
  library(edgeR)
  
  
  res.df<-as.data.frame(res)
  res.df$log10MeanNormCount<-log10(res.df$baseMean)
  
  # filter not expressed genes
  idx<-rowSums(counts(dds.diff))>5
  res.df.fil<-(res.df[idx,])
  
  res.df.fil$padj[is.na(res.df.fil$padj)]<-1
  
  #get stat
  res.sig<-res[which(res$padj<0.05 & abs(res$log2FoldChange)>=1),]
  wx<-which(rownames(res.df.fil) %in% rownames(res.sig))
  stat<-rep(0,nrow(res.df.fil))
  stat[wx]<-rep(1,length(wx))
  
  
  glMDPlot(res.df.fil, 
           xval = "log10MeanNormCount",
           yval="log2FoldChange",
           counts=counts(dds.diff, normalized=T)[idx,],
           anno=data.frame(GeneID=rownames(dds.diff)[idx]),
           groups=dds.diff$condition, samples=colnames(dds.diff),
           display.columns = c("symbol", "EnsemblID",
                               "biotype","pvalue","padj"),
           status = stat,
           path = path.diff.sh3)
  
##########################
save(res.sig, file=paste0(path.diff.sh3,"/res.sig.rda"))

## down and up -regulated genes
res.up<-subset(res.sig, log2FoldChange>0)
save(res.up, file=paste0(path.diff.sh3,"/res.up.rda"))
  
res.down<-subset(res.sig, log2FoldChange<0)
save(res.down, file=paste0(path.diff.sh3,"/res.down.rda"))
  
write.table(res.up$symbol, 
              file=paste0(path.diff.sh3,"/upReg_genes_symbol.txt"), sep="\n",
              quote=F, col.names=F, row.names = F)
write.table(res.down$symbol,
              file=paste0(path.diff.sh3,"/downReg_genes_symbol.txt"), sep="\n",
              quote=F, col.names=F, row.names = F)
  
write.table(res.sig[order(res.sig$padj),],
              file=paste0(path.diff.sh3,"/signif_genes.txt"), sep="\t",
              quote=F, col.names=T, row.names = F)


### heatmap

pdf(paste0(path.diff.sh3,"/heatmap_signif_genes_all_conditions.pdf"),
    height=1+(nrow(mat)/100), onefile = F)
    print(pheatmap(mat,scale="row", annotation_col = anno,show_rownames  = F) )
dev.off()
 
```

##Session info
```{r info}
sessionInfo()


```



